formats/json-tests/jvmTest/src/kotlinx/serialization/SerializerByTypeCacheTest.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2017-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.json.Json
+import java.net.URLClassLoader
+import kotlin.reflect.*
+import kotlin.test.*
+
+class SerializerByTypeCacheTest {
+
+    @Serializable
+    class Holder(val i: Int)
+
+    @Suppress("UNCHECKED_CAST")
+    @Test
+    fun testCaching() {
+        val typeOfKType = typeOf<Holder>()
+        val parameterKType = typeOf<List<Holder>>().arguments[0].type!!
+        assertSame(serializer(), serializer<Holder>())
+        assertSame(serializer(typeOfKType), serializer(typeOfKType))
+        assertSame(serializer(parameterKType), serializer(parameterKType))
+        assertSame(serializer(), serializer(typeOfKType) as KSerializer<Holder>)
+        assertSame(serializer(parameterKType) as KSerializer<Holder>, serializer(typeOfKType) as KSerializer<Holder>)
+    }
+
+    /**
+     * Checking the case when a parameterized type is loaded in different parallel [ClassLoader]s.
+     *
+     * If the main type is loaded by a common parent [ClassLoader] (for example, a bootstrap for [List]),
+     * and the element class is loaded by different loaders, then some implementations of the [KType] (e.g. `KTypeImpl` from reflection) may not see the difference between them.
+     *
+     * As a result, a serializer for another loader will be returned from the cache, and it will generate instances, when working with which we will get an [ClassCastException].
+     *
+     * The test checks the correctness of the cache for such cases - that different serializers for different loaders will be returned.
+     *
+     * [see](https://youtrack.jetbrains.com/issue/KT-54523).
+     */
+    @Test
+    fun testDifferentClassLoaders() {
+        val elementKType1 = SimpleKType(loadClass().kotlin)
```

Comment:
Is there any way to test this on real `KTypeImpl` instead of hand-rolled?

Reply:
In this case, it will be necessary to add a dependency on `kotlin-reflect` for all tests.